commit 9aa0a489d909af0cc36c41d3061ef956c7442ce2
Author: Roland Dreier <rolandd@cisco.com>
Date:   Sat Jun 13 15:14:09 2009 -0700

    IB/mthca: Don't double-free IRQs when falling back from MSI-X to INTx
    
    When both MSI-X and legacy INTx fail to generate an interrupt, the
    driver frees the MSI-X interrupts twice.  Fix this by clearing the
    have_irq flag for the MSI-X interrupts when they are freed the first
    time.
    
    Reported-by: Yinghai Lu <yhlu.kernel@gmail.com>
    Tested-by: Yinghai Lu <yhlu.kernel@gmail.com>
    Signed-off-by: Roland Dreier <rolandd@cisco.com>

diff --git a/drivers/infiniband/hw/mthca/mthca_eq.c b/drivers/infiniband/hw/mthca/mthca_eq.c
index 28f0e0c..90e4e45 100644
--- a/drivers/infiniband/hw/mthca/mthca_eq.c
+++ b/drivers/infiniband/hw/mthca/mthca_eq.c
@@ -641,9 +641,11 @@ static void mthca_free_irqs(struct mthca_dev *dev)
 	if (dev->eq_table.have_irq)
 		free_irq(dev->pdev->irq, dev);
 	for (i = 0; i < MTHCA_NUM_EQ; ++i)
-		if (dev->eq_table.eq[i].have_irq)
+		if (dev->eq_table.eq[i].have_irq) {
 			free_irq(dev->eq_table.eq[i].msi_x_vector,
 				 dev->eq_table.eq + i);
+			dev->eq_table.eq[i].have_irq = 0;
+		}
 }
 
 static int mthca_map_reg(struct mthca_dev *dev,
